Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp

```diff
@@ -0,0 +1,293 @@
+        self->m_platformSynthesizer.client().voicesDidChange();
+    }), this);
+
+    m_speakerCreatedCallback();
+}
+
+SpielSpeechWrapper::~SpielSpeechWrapper()
+{
+    g_signal_handlers_disconnect_by_data(m_speaker.get(), this);
+}
+
+String SpielSpeechWrapper::generateVoiceURI(const GRefPtr<SpielVoice>& voice, const String& language)
+{
+    auto provider = spiel_voice_get_provider(voice.get());
+    return makeString(URI_PREFIX, span(spiel_provider_get_well_known_name(provider)), '#', span(spiel_voice_get_identifier(voice.get())), '#', language);
+}
+
+Vector<RefPtr<PlatformSpeechSynthesisVoice>> SpielSpeechWrapper::initializeVoiceList()
+{
+    Vector<RefPtr<PlatformSpeechSynthesisVoice>> platformVoices;
+    auto voices = spiel_speaker_get_voices(m_speaker.get());
+    unsigned position = 0;
+    m_voices.clear();
+    while (auto item = g_list_model_get_item(voices, position)) {
+        auto voice = SPIEL_VOICE(item);
+        auto name = makeString(span(spiel_voice_get_name(voice)));
+        const char* const* languages = spiel_voice_get_languages(voice);
+        for (unsigned i = 0; i < G_N_ELEMENTS(languages); i++) {
+            auto language = makeString(span(languages[i]));
+            bool isDefault = !i;
+            auto uri = generateVoiceURI(voice, language);
+            platformVoices.append(PlatformSpeechSynthesisVoice::create(uri, name, language, false, isDefault));
+            m_voices.add(uri, GRefPtr(voice));
+        }
+        position++;
+    }
+    return platformVoices;
+}
+
+void SpielSpeechWrapper::pause()
+{
+    if (!m_utterance)
+        return;
+
+    spiel_speaker_pause(m_speaker.get());
+}
+
+void SpielSpeechWrapper::resume()
+{
+    if (!m_utterance)
+        return;
+
+    spiel_speaker_resume(m_speaker.get());
+}
+
+void SpielSpeechWrapper::speakUtterance(RefPtr<PlatformSpeechSynthesisUtterance>&& utterance)
+{
+    ASSERT(!m_utterance);
+    ASSERT(utterance);
+    if (!utterance)
+        return;
+
+    if (!utterance->voice()) {
+        m_platformSynthesizer.client().didFinishSpeaking(*utterance);
+        return;
+    }
+
+    const auto& uri = utterance->voice()->voiceURI();
+    if (!m_voices.contains(uri)) {
+        WTFLogAlways("Unknown voice URI: %s", uri.utf8().data());
+        return;
+    }
+
+    auto voice = m_voices.get(uri);
+    auto text = utterance->text().utf8();
+    auto spielUtterance = adoptGRef(spiel_utterance_new(text.data()));
+    auto language = utterance->lang().utf8();
+    spiel_utterance_set_language(spielUtterance.get(), language.data());
+    spiel_utterance_set_voice(spielUtterance.get(), voice);
+    spiel_utterance_set_volume(spielUtterance.get(), utterance->volume());
+    spiel_utterance_set_pitch(spielUtterance.get(), utterance->pitch());
+    spiel_utterance_set_rate(spielUtterance.get(), utterance->rate());
+    m_utterance = WTFMove(utterance);
+    spiel_speaker_speak(m_speaker.get(), spielUtterance.get());
```

Comment:
Is there a chance m_speaker is null here if speak is called before the async init is done? Looking at this on my phone so maybe I missed something :)

If init isn't done the speak() needs to be deferred.

Reply:
Yep, I believe this is handled, scroll down a bit to `PlatformSpeechSynthesizer::speak()` ;)

Reply:
😎 Thanks for working on this!